consensus/service/src/byzantine_ledger/worker.rs

```diff
@@ -418,183 +404,209 @@ impl<
     }
 
-    fn process_consensus_msgs_for_cur_slot(&mut self) {
-        if let Some(consensus_msgs) = self.pending_consensus_msgs.remove(&self.cur_slot) {
-            for (consensus_msg, from_responder_id) in consensus_msgs {
-                let (scp_msg, their_prev_block_id) =
-                    (consensus_msg.scp_msg(), consensus_msg.prev_block_id());
-
-                if self.prev_block_id != *their_prev_block_id {
-                    log::warn!(self.logger, "Received message {:?} that refers to an invalid block id {:?} != {:?} (cur_slot = {})",
-                    scp_msg, their_prev_block_id, self.prev_block_id, self.cur_slot);
-                }
+    // Process messages for current slot and recent previous slots; retain messages for future slots.
+    fn process_consensus_msgs(&mut self) {
+        // Process messages for slot indices in [oldest_slot, current_slot].
+        let current_slot = self.current_slot_index;
+        let max_externalized_slots = self.scp.max_externalized_slots() as u64;
+        let oldest_slot = current_slot.saturating_sub(max_externalized_slots);
+        let (consensus_msgs, future_msgs): (Vec<_>, Vec<_>) = self
+            .pending_consensus_msgs
+            .drain(..)
+            // We do not perform consensus on the origin block.
+            .filter(|(consensus_msg, _)| consensus_msg.scp_msg().slot_index != 0)
+            .filter(|(consensus_msg, _)| consensus_msg.scp_msg().slot_index >= oldest_slot)
+            .partition(|(consensus_msg, _)| consensus_msg.scp_msg().slot_index <= current_slot);
+
+        self.pending_consensus_msgs = future_msgs;
+
+        // Omit "incompatible" messages that refer to a different blockchain history.
+        let (compatible_msgs, incompatible_msgs): (Vec<_>, Vec<_>) =
+            consensus_msgs.into_iter().partition(|(consensus_msg, _)| {
+                let previous_block_id = self
+                    .ledger
+                    .get_block(consensus_msg.scp_msg().slot_index - 1)
+                    .expect("This block must be in our ledger")
+                    .id;
+                *consensus_msg.prev_block_id() == previous_block_id
+            });
+
+        for (consensus_msg, _) in &incompatible_msgs {
+            log::warn!(
+                self.logger,
+                "Msg refers to a different blockchain. Msg {}, previous block ID: {:?}",
+                consensus_msg.scp_msg().to_display(),
+                consensus_msg.prev_block_id(),
+            );
+        }
 
-                if !self.fetch_missing_txs(scp_msg, &from_responder_id) {
-                    continue;
-                }
+        // Process compatible messages in batches.
+        let batch_size = 5;
+        for chunk in compatible_msgs.chunks(batch_size) {
+            // Omit a message if it references a transaction that cannot be obtained.
+            let (resolved, failed): (Vec<_>, Vec<_>) =
+                chunk
+                    .iter()
```

Comment:
I wonder how this would behave if we used a parallel iterator (https://docs.rs/rayon/0.6.0/rayon/par_iter/index.html)
(out of scope for this PR but would be interesting to try it out)